### ISSUE TITLE
Statically access Config does not decrypt values

### ISSUE THREAD

**Issue by ismagilov** [svenkubiak/mangooio#282]
While I was playing with "encrypted values", I tried to encrypt `application.admin.username`. Here is a relevant part of a config:
```
dev:
    application:
        admin:
            enable: true
            username: cryptex[XnrslCZdmq9BW27skRxH5g==] #ilya - encrypted with a key from pwd.txt: sderokfoksm458349034rkwelkw34njk
            password: "$2a$06$OCwDt.egSbum93F9DGgOo.5Fqgl3Je.IrV1HeELOMTtGZ87YDJ2Xm" #12345 - JBcrypt
        masterkey: /Users/ilya/Work/Projects/todobackend-mangooio/pwd.txt # contains key: sderokfoksm458349034rkwelkw34njk
```

Here I used encrypted `username`: `cryptex[...]`. After I had done that, I couldn`t login into administration page anymore. Admin page displayed nothing. The page was reloaded and login dialog appeared again. Logs were empty.

**Comment by svenkubiak** [comment]
Why did you put the pasword in Qoutes? Nevertheless, will look into it.

**Comment by ismagilov** [comment]
> Why did you put the pasword in Qoutes?

I couldn't setup admin access at once. I used this online service: https://bcrypt-generator.com. It generated a hash, but it didn't work. I thought, it was because of "special" symbols, like `$.`, that's why I put the password in quotes. Then I tried another online service: http://bcrypthashgenerator.apphb.com. It's hash worked. Finally, I forgot to remove them.

As soon as online generators are not reliable, and there is no way to access admin tools without having hashed password, would it be better:
* to have some standalone util for generating hashes for passwords?
* to let login to admin without a password (then you can generate hash using built-in tool)?
* to skip passwords hashing, and let users encrypt their passwords using master key?

For me all this three "security" realted topics ("application secret"/"password hashing with `bcrypt`"/"master password for values") looked a little bit messy (or complex) when I read docs. Probably, it could be a subject for simplification or more clear explanation.

**Comment by svenkubiak** [comment]
Typicall use case (at leas IMHO) would be to access admin interface without authentication in dev mode and then create the credentials for prod environment. But I agree, having JBcrypt and Encryption in the config file could be kind of misleading. What do you think about the following idea:

Remove JBcrypt requirement from config file and instead just use cryptext[....] encryption. Then one could use encryption outside the framework and pass the value to the config file. 

What do you think?

**Comment by ismagilov** [comment]
> Typicall use case (at least IMHO) would be to access admin interface without authentication in dev mode and then create the credentials for prod environment.

Is it how it works now? I tried to skip `password` and couldn't login. 
```
dev:
    application:
        admin:
            enable: true
            username: ilya
```

> Remove JBcrypt requirement from config file and instead just use cryptext[....] encryption. Then one could use encryption outside the framework and pass the value to the config file.

1. I think that a framework itself should provide encryption tools. I have looked through *Play*, they have such tools. So the idea to give access to admin tools in dev mode without password and let user get encrypted strings is great.

2. Using hashed passwords has one advantage: password are not saved in memory in plain text. It protects you from `memory dumps` attacks, but I am not confident with that, as I am not security professional... Except application password, there can be passwords to db, external services, etc. And you can't use hashes for them. So it would be more simple to use one way to encrypt passwords throughout a config. So here I agree with you, JBcrypt is a little bit redundant from this point of view.

3. Application secret. I would recommend you to describe more concretely for what purposes it is used, and why it is bad idea to publish it in VCS. [Play describes it very well](https://www.playframework.com/documentation/2.5.x/ApplicationSecret). They give an opportunity to pass secret though command line in production. In your framework you support config values encryption. So probably it will be reasonable either to force its encryption, or to state explicitly in docs that it must be encrypted.

P.S. And what about the main issue? I saw that you marked the issue as not being a bug. I fixed a typo in config (`cryptex[...]` - last 't' was missing), but it stills doesn't work.

**Comment by svenkubiak** [comment]
> P.S. And what about the main issue? I saw that you marked the issue as not being a bug. I fixed a typo in config (cryptex[...] - last 't' was missing), but it stills doesn't work.

Please try to create the hashed JBCrypt Password in the Admin Tools, leaving the username as cleartext in the config. If this does **not** work, we have a bad unit test. If this **does** work, create the encrypted username also with the Admin Tools using your masterkey. Then check, if this works.

> Is it how it works now? I tried to skip password and couldn't login.

You need username AND password. Otherwise it won't work.

> 1. I think that a framework itself should provide encryption tools. I have looked through Play, they have such tools. So the idea to give access to admin tools in dev mode without password and let user get encrypted strings is great.

Yes, Play has a lot of Python- and Shell Script aresenal for doing stuff. I am not sure, if it is a good way to go down that road.

> 2. Using hashed passwords has one advantage: password are not saved in memory in plain text. It protects you from memory dumps attacks, but I am not confident with that, as I am not security professional... Except application password, there can be passwords to db, external services, etc. And you can't use hashes for them. So it would be more simple to use one way to encrypt passwords throughout a config. So here I agree with you, JBcrypt is a little bit redundant from this point of view.

Either have plaintext OR encrypted values, sounds like a more simple, yet good approach on that. Agree.

> 3. Application secret. I would recommend you to describe more concretely for what purposes it is used, and why it is bad idea to publish it in VCS. Play describes it very well. They give an opportunity to pass secret though command line in production. In your framework you support config values encryption. So probably it will be reasonable either to force its encryption, or to state explicitly in docs that it must be encrypted.

Agree on enhancing documentation on this.

**Comment by ismagilov** [comment]
> Please try to create the hashed JBCrypt Password in the Admin Tools, leaving the username as cleartext in the config. If this does not work, we have a bad unit test.

It works.

```
username: ilya 
password: $2a$12$dLSHOqKzDEt6tictTGbn8epV0kTEbiWpHx/VVRDoUVpFiS2bu6I6y #password: 12345 encrypted by Admin tool
```

> If this does work, create the encrypted username also with the Admin Tools using your masterkey. Then check, if this works.

It doesn't work.

```
username: cryptex[XnrslCZdmq9BW27skRxH5g==] #encrypted: ilya with master key: sderokfoksm458349034rkwelkw34njk
password: $2a$12$dLSHOqKzDEt6tictTGbn8epV0kTEbiWpHx/VVRDoUVpFiS2bu6I6y #password: 12345 encrypted by Admin tool
```

**Comment by svenkubiak** [comment]
Can you test it without the masterkey and using the application secret for the encryption? Sorry for asking, but trying to drill down the possible bug.

**Comment by ismagilov** [comment]
But how can I test without master key? Doesn't framework demand it, if I use `cryptex[...]` in a config?

I tried and got:
```
Exception in thread "main" java.lang.NullPointerException
	at java.io.File.<init>(File.java:277)
	at io.mangoo.configuration.Config.getMasterKey(Config.java:153)
	at io.mangoo.configuration.Config.decrypt(Config.java:129)
	at io.mangoo.core.Application.main(Application.java:52)
```

Well, seems that I didn't understand the case you are asking to test...

**Comment by svenkubiak** [comment]
Yes, you are write. Found the bug meanwhile :-)

**Comment by svenkubiak** [comment]
Fixed in 4.3.1
